File: android/app/src/main/java/com/gediao9/app/AudioRecorderPlugin.java
package com.gediao9.app;

import android.Manifest;
import android.content.Intent;
import android.media.MediaRecorder;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;

import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import org.json.JSONArray;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

@CapacitorPlugin(
    name = "AudioRecorder",
    permissions = {
        @Permission(alias = "microphone", strings = { Manifest.permission.RECORD_AUDIO })
    }
)
public class AudioRecorderPlugin extends Plugin {
    private MediaRecorder mediaRecorder;
    private Handler segmentHandler;
    private Runnable segmentRunnable;
    private final List<JSObject> segments = new ArrayList<>();
    private boolean isRecording = false;
    private long segmentInterval = 30000L;
    private int currentSegmentIndex = 0;
    private long currentSegmentStartTime = 0L;
    private File currentSegmentFile;

    @PluginMethod
    public void startRecording(PluginCall call) {
        if (getPermissionState("microphone") != PermissionState.GRANTED) {
            requestPermissionForAlias("microphone", call, "startRecordingAfterPermission");
            return;
        }

        startRecordingInternal(call);
    }

    @PermissionCallback
    private void startRecordingAfterPermission(PluginCall call) {
        if (getPermissionState("microphone") == PermissionState.GRANTED) {
            startRecordingInternal(call);
        } else {
            call.reject("Microphone permission denied");
        }
    }

    private void startRecordingInternal(PluginCall call) {
        if (isRecording) {
            cleanupState();
            call.reject("Recording already in progress");
            return;
        }

        cleanupState();

        Long interval = call.getLong("segmentInterval", 30000L);
        segmentInterval = interval == null ? 30000L : interval;
        currentSegmentIndex = 0;
        segments.clear();
        segmentHandler = new Handler(Looper.getMainLooper());
        isRecording = true;

        Intent serviceIntent = new Intent(getContext(), RecordingForegroundService.class);
        serviceIntent.setAction(RecordingForegroundService.ACTION_START);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            getContext().startForegroundService(serviceIntent);
        } else {
            getContext().startService(serviceIntent);
        }

        try {
            startCurrentSegment();
            call.resolve();
        } catch (Exception error) {
            isRecording = false;
            cleanupState();
            stopForegroundService();
            notifyError("Failed to start recording: " + error.getMessage());
            call.reject("Failed to start recording: " + error.getMessage());
        }
    }

    @PluginMethod
    public void stopRecording(PluginCall call) {
        if (!isRecording) {
            call.reject("Not recording");
            return;
        }

        isRecording = false;

        if (segmentHandler != null) {
            segmentHandler.removeCallbacksAndMessages(null);
            segmentHandler = null;
        }
        segmentRunnable = null;

        stopCurrentSegment(true);
        stopForegroundService();

        JSObject result = new JSObject();
        JSONArray segmentArray = new JSONArray();
        for (JSObject segment : segments) {
            segmentArray.put(segment);
        }
        result.put("segments", segmentArray);
        segments.clear();
        call.resolve(result);
    }

    @PluginMethod
    public void getSegments(PluginCall call) {
        JSObject result = new JSObject();
        JSONArray segmentArray = new JSONArray();
        for (JSObject segment : segments) {
            segmentArray.put(segment);
        }
        result.put("segments", segmentArray);
        call.resolve(result);
    }

    private void startCurrentSegment() throws Exception {
        currentSegmentFile = new File(
            getContext().getCacheDir(),
            "segment_" + System.currentTimeMillis() + "_" + currentSegmentIndex + ".m4a"
        );

        mediaRecorder = createRecorder();
        mediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
        mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
        mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);
        mediaRecorder.setAudioSamplingRate(16000);
        mediaRecorder.setAudioEncodingBitRate(64000);
        mediaRecorder.setAudioChannels(1);
        mediaRecorder.setOutputFile(currentSegmentFile.getAbsolutePath());
        mediaRecorder.prepare();
        mediaRecorder.start();
        currentSegmentStartTime = System.currentTimeMillis();

        segmentRunnable = () -> {
            if (!isRecording) return;
            rotateSegment();
        };
        segmentHandler.postDelayed(segmentRunnable, segmentInterval);
    }

    private MediaRecorder createRecorder() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return new MediaRecorder(getContext());
        }

        return new MediaRecorder();
    }

    private void rotateSegment() {
        stopCurrentSegment(true);
        currentSegmentIndex += 1;

        if (!isRecording) return;

        try {
            startCurrentSegment();
        } catch (Exception error) {
            isRecording = false;
            if (segmentHandler != null) {
                segmentHandler.removeCallbacksAndMessages(null);
                segmentHandler = null;
            }
            segmentRunnable = null;
            stopForegroundService();
            notifyError("Failed to rotate recording segment: " + error.getMessage());
        }
    }

    private void stopCurrentSegment(boolean emitSegment) {
        if (mediaRecorder != null) {
            try {
                mediaRecorder.stop();
            } catch (RuntimeException ignored) {
                // Android throws if a segment is too short or contains no audio data.
            }
            mediaRecorder.release();
            mediaRecorder = null;
        }

        if (!emitSegment || currentSegmentFile == null || !currentSegmentFile.exists() || currentSegmentFile.length() == 0) {
            return;
        }

        double duration = (System.currentTimeMillis() - currentSegmentStartTime) / 1000.0;
        JSObject segment = new JSObject();
        segment.put("index", currentSegmentIndex);
        segment.put("filePath", currentSegmentFile.getName());
        segment.put("duration", duration);
        segment.put("size", currentSegmentFile.length());
        segments.add(segment);
        notifyListeners("segmentAvailable", segment);
    }

    private void notifyError(String message) {
        JSObject error = new JSObject();
        error.put("message", message);
        notifyListeners("recordingError", error);
    }

    private void cleanupState() {
        if (mediaRecorder != null) {
            try {
                mediaRecorder.stop();
            } catch (RuntimeException ignored) {
            }
            mediaRecorder.release();
            mediaRecorder = null;
        }

        if (segmentHandler != null) {
            segmentHandler.removeCallbacksAndMessages(null);
            segmentHandler = null;
        }
        segmentRunnable = null;
        isRecording = false;
    }

    private void stopForegroundService() {
        Intent serviceIntent = new Intent(getContext(), RecordingForegroundService.class);
        serviceIntent.setAction(RecordingForegroundService.ACTION_STOP);
        getContext().startService(serviceIntent);
    }
}

File: capacitor-plugins/audio-recorder/android/src/main/java/com/gediao9/audiorecorder/AudioRecorderPlugin.kt
package com.gediao9.audiorecorder

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.media.MediaRecorder
import android.os.Build
import android.os.Handler
import android.os.Looper
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import java.io.File
import java.io.IOException

@CapacitorPlugin(
    name = "AudioRecorder",
    permissions = [
        Permission(
            alias = "microphone",
            strings = [Manifest.permission.RECORD_AUDIO]
        )
    ]
)
class AudioRecorderPlugin : Plugin() {

    private var mediaRecorder: MediaRecorder? = null
    private var segmentHandler: Handler? = null
    private var segmentRunnable: Runnable? = null
    private var segmentIndex = 0
    private var isRecording = false
    private var segmentInterval: Long = 30000
    private var tempDir: File? = null
    private var segments = mutableListOf<Map<String, Any>>()
    private var startTime: Long = 0

    companion object {
        private const val NOTIFICATION_CHANNEL_ID = "audio_recorder_channel"
        private const val NOTIFICATION_ID = 1
    }

    @PluginMethod
    fun startRecording(call: PluginCall) {
        if (getPermissionState("microphone") != PermissionState.GRANTED) {
            requestPermissionForAlias("microphone", call, "startRecordingAfterPermission")
            return
        }
        executeStartRecording(call)
    }

    private fun startRecordingAfterPermission(call: PluginCall) {
        if (getPermissionState("microphone") == PermissionState.GRANTED) {
            executeStartRecording(call)
        } else {
            call.reject("Microphone permission denied")
        }
    }

    private fun executeStartRecording(call: PluginCall) {
        if (isRecording) {
            call.reject("Recording already in progress")
            return
        }

        segmentInterval = call.getLong("segmentInterval", 30000)
        segmentIndex = 0
        segments.clear()
        tempDir = context.cacheDir

        segmentHandler = Handler(Looper.getMainLooper())

        try {
            createNotificationChannel()
            startForegroundNotification()
            startNextSegment()
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to start recording: ${e.message}")
        }
    }

    @PluginMethod
    fun stopRecording(call: PluginCall) {
        if (!isRecording) {
            call.reject("Not recording")
            return
        }

        isRecording = false
        segmentRunnable?.let { segmentHandler?.removeCallbacks(it) }
        mediaRecorder?.stop()
        mediaRecorder?.release()
        mediaRecorder = null
        stopForegroundNotification()

        val result = JSObject()
        val segmentsArray = JSObject()
        for ((i, segment) in segments.withIndex()) {
            val segObj = JSObject()
            segObj.put("index", segment["index"] as Int)
            segObj.put("filePath", segment["filePath"] as String)
            segObj.put("duration", segment["duration"] as Double)
            segObj.put("size", segment["size"] as Long)
            segmentsArray.put(i.toString(), segObj)
        }
        result.put("segments", segmentsArray)
        segments.clear()
        call.resolve(result)
    }

    @PluginMethod
    fun getSegments(call: PluginCall) {
        val result = JSObject()
        val segmentsArray = JSObject()
        for ((i, segment) in segments.withIndex()) {
            val segObj = JSObject()
            segObj.put("index", segment["index"] as Int)
            segObj.put("filePath", segment["filePath"] as String)
            segObj.put("duration", segment["duration"] as Double)
            segObj.put("size", segment["size"] as Long)
            segmentsArray.put(i.toString(), segObj)
        }
        result.put("segments", segmentsArray)
        call.resolve(result)
    }

    private fun startNextSegment() {
        if (!isRecording && segmentIndex > 0) return

        mediaRecorder?.stop()
        mediaRecorder?.release()

        val outputFile = File(tempDir, "segment_$segmentIndex.m4a")
        val duration = (System.currentTimeMillis() - startTime) / 1000.0
        val fileSize = outputFile.length()

        if (segmentIndex > 0 || isRecording) {
            val segment = mapOf(
                "index" to segmentIndex,
                "filePath" to outputFile.absolutePath,
                "duration" to duration,
                "size" to fileSize
            )
            segments.add(segment)

            val data = JSObject()
            data.put("index", segmentIndex)
            data.put("filePath", outputFile.absolutePath)
            data.put("duration", duration)
            data.put("size", fileSize)
            notifyListeners("segmentAvailable", data)
        }

        isRecording = true
        segmentIndex++

        mediaRecorder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }.apply {
            setAudioSource(MediaRecorder.AudioSource.MIC)
            setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            setAudioSamplingRate(16000)
            setAudioChannels(1)
            setOutputFile(outputFile.absolutePath)
            try {
                prepare()
                start()
                startTime = System.currentTimeMillis()
            } catch (e: IOException) {
                val error = JSObject()
                error.put("message", "Failed to prepare recorder: ${e.message}")
                notifyListeners("recordingError", error)
                return
            }
        }

        segmentRunnable = Runnable {
            startNextSegment()
        }
        segmentHandler?.postDelayed(segmentRunnable!!, segmentInterval)
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                NOTIFICATION_CHANNEL_ID,
                "录音服务",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = "保持录音在后台运行"
            }
            val manager = context.getSystemService(NotificationManager::class.java)
            manager.createNotificationChannel(channel)
        }
    }

    private fun startForegroundNotification() {
        val notification = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Notification.Builder(context, NOTIFICATION_CHANNEL_ID)
        } else {
            @Suppress("DEPRECATION")
            Notification.Builder(context)
        }.apply {
            setContentTitle("格调9问")
            setContentText("正在录音...")
            setSmallIcon(android.R.drawable.ic_btn_speak_now)
            setOngoing(true)
        }.build()

        // Note: In a real app, you'd need to use startForeground() from a Service
        // This is a simplified version
    }

    private fun stopForegroundNotification() {
        val manager = context.getSystemService(NotificationManager::class.java)
        manager.cancel(NOTIFICATION_ID)
    }
}
